python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py

```diff
@@ -81,26 +83,9 @@ def post_training_quantize(self, model, calib_dataloader=None, val_dataloader=No
         if val_dataloader:
             self.eval_dataloader = val_dataloader
             if metric:
-                framework = self.cfg.model.framework
-                if 'pytorch' in framework:
-                    framework_metric = METRICS['pytorch']
-                else:
-                    framework_metric = METRICS[framework]
-
-                class MyMetric(framework_metric):
-                    def __init__(self):
-                        """
-                        This local class is to resolve dumping issue in tensorflow.
-                        In tensorflow, INC will try to dump the metric to yaml which
-                        somehow causes unexpected error. So we moved metric assignment
-                        to the new local class to avoid that.
-                        """
-                        self.metric = metric
-
-                self.metric = Metric(
-                    MyMetric,
-                    name="%s_%s" % (framework_metric.__name__, type(metric).__name__)
-                )
+                framework_inc_metric = INCMETRICS[self.cfg.model.framework]
+                framework_inc_metric.set_metric(metric)
```

Comment:
Does it make sense to modify a global state in a object? I think it is hard reason about the behavior and maintain the code. E.g. If I have a second `QuantizationINC` object with different metric, will it mess up the state of the first one?

Reply:
You can't run them in parallel. Otherwise, this mess would be possible.

Reply:
If you don't use multiple threads to do twice quantization in your assumption, that'll be fine. And seems doing that is not a good practice for post-training quantization.

Reply:
I do not think you need multi-threading to mess it up. Does the following code work?
```python
quantA = QuantizationINC(metric-a)
quantB = QuantizationINC(metric-b)

quantA.some_method
quantB.some_method
```

Reply:
If there is a serialization problem, I think you should try to solve the serialization problem directly.

Reply:
> I do not think you need multi-threading to mess it up. Does the following code work?
> 
> ```python
> quantA = QuantizationINC(metric-a)
> quantB = QuantizationINC(metric-b)
> 
> quantA.some_method
> quantB.some_method
> ```
Your use case is not actually correct. It should be
```python
trainer.quantize(model, trainloader, valloader, metric_a)
trainer.quantize(model, trainloader, valloader, metric_b)
```
or
```python
trainer.quantize(model, trainloader, valloader, metric_a)
trainer.quantize(model, trainloader, valloader, metric_b)
```python
quantA = QuantizationINC(**kwargs_a)
quantB = QuantizationINC(**kwargs_b)

quantized_1 = quantA.post_training_quantize(model, calib_dataloader, val_dataloader,
                                             metric_a)
quantized_2 = quantB.post_training_quantize(model, calib_dataloader, val_dataloader,
                                             metric_b)
quantized_3 = quantA.post_training_quantize(model, calib_dataloader, val_dataloader,
                                             metric_b)
quantized_4 = quantB.post_training_quantize(model, calib_dataloader, val_dataloader,
                                             metric_a)
```
With different kwargs and metric, the 4 quantization processes are all different and the metric wouldn't be messed up.

Reply:
> If there is a serialization problem, I think you should try to solve the serialization problem directly.

If we want to resolve this directly, then metric pass to INC must be `{cls: kwargs}` and we need to assure that kwargs must be able to be serialized. So the solution is to pass a dict `{torchmetric_class: kwargs}` instead of instance of `torchmetric_class(kwargs)` as stated in comments below. Having discussed with @TheaperDeng , we agreed on passing a torchmetric instance which is more user-friendly.

Reply:
> If there is a serialization problem, I think you should try to solve the serialization problem directly.

Or we need to check in some changes into INC directly... But this is not possible for the moment.